src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnection.cs

```diff
@@ -143,6 +149,73 @@ private Task OnConnectedAsyncCore(OpenConnectionMessage message)
             }
         }
 
+        private void ProcessOutgoingMessages(ConnectionDataMessage connectionDataMessage)
+        {
+            var connectionId = connectionDataMessage.ConnectionId;
+            if (_clientConnections.TryGetValue(connectionId, out var transport))
+            {
+                try
+                {
+                    var payload = connectionDataMessage.Payload;
+                    Log.WriteMessageToApplication(_logger, payload.Length, connectionId);
+                    var message = GetString(payload);
+                    if (message == ReconnectMessage)
+                    {
+                        transport.Reconnected?.Invoke();
+                    }
+                    else
+                    {
+                        transport.OnReceived(message);
+                    }
+                }
+                catch (Exception e)
+                {
+                    Log.FailToWriteMessageToApplication(_logger, connectionDataMessage.ConnectionId, e);
+                }
+            }
+            else
+            {
+                // Unexpected error
+                Log.ReceivedMessageForNonExistentConnection(_logger, connectionDataMessage.ConnectionId);
+            }
+        }
+
+        private async Task ProcessMessageAsync(string connectionId)
+        {
+            // Check if channel is created.
+            if (_clientChannels.TryGetValue(connectionId, out var channel))
+            {
+                try
+                {
+                    // Check if channel is closed.
+                    while (await channel.Reader.WaitToReadAsync())
+                    {
+                        while (channel.Reader.TryRead(out var serviceMessage))
+                        {
+                            switch (serviceMessage)
+                            {
+                                case OpenConnectionMessage openConnectionMessage:
+                                    await OnConnectedAsyncCore(openConnectionMessage);
+                                    break;
+                                case CloseConnectionMessage closeConnectionMessage:
+                                    PerformDisconnectCore(closeConnectionMessage.ConnectionId);
+                                    break;
+                                case ConnectionDataMessage connectionDataMessage:
+                                    ProcessOutgoingMessages(connectionDataMessage);
+                                    break;
+                                default:
+                                    break;
+                            }
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    Log.SendLoopStopped(_logger, connectionId, e);
```

Comment:
If loop stopped, should we stop the channel to notify writer?

Reply:
PerformDisconnect to clean and create a disconnect message to notify service

---
In reply to: [237335224](https://github.com/Azure/azure-signalr/pull/268#discussion_r237335224) [](ancestors = 237335224)